### PR TITLE
Consider None as return value from pre-processor rather than assume mutation

### LINKED ISSUE #347: Unenveloping None values doesn't work as expected
Trying to unenvelope serialized data when the contained data is None continues deserialization with the original data. This makes it impossible to deserialize nested fields with enveloped data where None is a valid value. (Related: http://jsonapi.org/format/)

Example (let's pretend our system allows for homeless employees):
https://gist.github.com/tdevelioglu/8e433355494ba022f48f
(I expect a deserialized Resource object, with an attribute 'address' with value 'None')

### PR DESCRIPTION
**Pull request by lafrech**
Fixes #347

TODO:

- [x] Update doc (incl. upgrading.rst)

### CODE REVIEW

File: tests/test_decorators.py
Comment (lafrech) [inline tests/test_decorators.py:111]: I feel weird about this test.

Returning `None` from ~`pre_dump`~`pre_load` does not make much sense.

Comment (sloria) [reply tests/test_decorators.py:111]: Do you mean `pre_load`? 

Agreed that it doesn't make sense for users to do that, but I think this is the expected behavior.

Comment (lafrech) [reply tests/test_decorators.py:111]: Yes, `pre_load`, sorry.

The test could be catching a `ValidationError` for some other reason, but there is no way to narrow it without testing private interface, so we can keep it this way.

**Review by sloria — APPROVED** [review]
Looks good to me. Once docs are updated, this is good to merge.

**Comment by lafrech** [comment]
Doc updated and PR rebased.

File: CHANGELOG.rst
Comment (sloria) [inline CHANGELOG.rst:9]: Thanks for making these consistent.

File: docs/upgrading.rst
Comment (sloria) [inline docs/upgrading.rst:465]: Would be nice to have a code example of how to update existing code, but that doesn't have to block merging of this PR.

**Review by sloria — APPROVED** [review]
Looks good. Thanks a lot for updating the changelog and upgrading guide.

File: CHANGELOG.rst
Comment (sloria) [inline CHANGELOG.rst:25]: This should be in 3.0.0b12. I'll fix this.

Comment (lafrech) [reply CHANGELOG.rst:25]: Right. I think it got screwed up during the rebase and I didn't notice.

**Pull request merged**
